### DISCUSSION TITLE
Preview action on domain object composed of viewable domain objects

### DISCUSSION THREAD

**Discussion by jordan-bonecutter** [nasa/openmct#5180] [Community Q&A] [answered]
I'm working on moving to the new(er) API in version 1. In the old API, if I wanted to preview a domain object representing a view which was composed of other domain obejcts, I could do the following:

```js
openmct.$injector.get('objectService').getObjects([key])
  .then((objs) => {
    const preview = objs[key].getCapability('action').getActions({ key: 'mct-preview-action' })[0]
    action.perform()
 })
```

With the new API, I'm assuming I would do something like:

```js
const obj = await openmct.objects.get(key)
const actions = await openmct.actions.get([obj])
actions.applicableActions.preview.callBack()
```

But this code only works for some domain objects, and I'm really not sure why. When it doesn't work I get the following errors in the console:

```
vue.min.js:6 TypeError: Cannot read property 'filter' of undefined
    at selectedLayoutItems (DisplayLayout.vue:169)
    at or.get (vue.min.js:6)
    at or.evaluate (vue.min.js:6)
    at s.selectedLayoutItems (vue.min.js:6)
    at s.<anonymous> (DisplayLayout.vue?a8b1:2)
    at s.e._render (vue.min.js:6)
    at s.i (vue.min.js:6)
    at or.get (vue.min.js:6)
    at new or (vue.min.js:6)
    at vue.min.js:6
vue.min.js:6 TypeError: Cannot read property 'forEach' of undefined
    at s.initializeItems (DisplayLayout.vue:482)
    at s.n [as initializeItems] (vue.min.js:6)
    at s.mounted (DisplayLayout.vue:207)
    at fe (vue.min.js:6)
    at Object.insert (vue.min.js:6)
    at w (vue.min.js:6)
    at We.__patch__ (vue.min.js:6)
    at We.e._update (vue.min.js:6)
    at We.i (vue.min.js:6)
    at or.get (vue.min.js:6)
```

I suspect it's because I'm not providing a view but I'm having trouble finding documentation on this feature. Any ideas?

**Comment by jvigliotta** [comment, marked as answer]
Hi @jordan-bonecutter! 

I believe you could do something like this:

``` js
import PreviewAction fom "@/ui/preview/PreviewAction.js";

const previewAction = new PreviewAction(openmct);
const objectPath = await openmct.objects.getOriginalPath(domainObject.identifier);
previewAction.invoke(objectPath);

````

We're going to be updating PreviewAction soon so that it will reuse any components that are already in the DOM (to enhance performance), but that should be a backward compatible change.

**Reply by jordan-bonecutter** [reply]
Thanks for the swift response!

Unfortunately my project isn't using typescript so I'm not exactly sure if I can wrangle this solution to work for me but I'll give it a whirl.

EDIT: don't think I need TS but `import` may not be an option, witout a significant code change at least.

**Reply by akhenry** [reply]
You can do also it via API without needing to include source (eg if you're building an app with Open MCT as a dependency) -
``` JavaScript
const objectPath = await openmct.objects.getOriginalPath(key);
const previewAction = openmct.actions.getAction('preview');

previewAction.invoke(objectPath);
```

**Reply by jordan-bonecutter** [reply]
The version my project uses lacks the `getAction` method. I'll attempt updating.

**Reply by akhenry** [reply]
Ah, ok, that was changed in an update last year. As a temporary workaround to get you unblocked you could try:
``` JavaScript
const objectPath = await openmct.objects.getOriginalPath(key);
const previewAction = openmct.actions._allActions['preview'];

previewAction.invoke(objectPath);
```

That's obviously relying on a private member variable and will break at some point, but hopefully it will get you unstuck until you can upgrade.

**Reply by jordan-bonecutter** [reply]
That worked! I'll add a note, we'll hopefully be upgrading to v2 soon. Thanks for all the help!

**Reply by jordan-bonecutter** [reply]
Out of curisosity, what was wrong with the code I provided above (if it's obvious)

**Reply by jvigliotta** [reply]
@jordan-bonecutter off the top of my head, not sure exactly why it was working for some objects and not others, but we've moved away from Angular injector and it's considered legacy code at this point (along with `getCapability` and `objectService`), we've removed that functionality from the latest releases. I believe we have some guidance on how to handle some of the more common "gotcha's" when upgrading from Legacy code, I will look for that and see if we can get a post up on the Discussions board with more information.
